### PR TITLE
DDP-4703 support special vars in activity templating

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

Add support for substituting "special vars" in templates, so we can pull in participant's name into activity content.

THIS DOES NOT SUPPORT ALL FORMS OF TEMPLATING. Only templates within an activity are supported, and only if they hook into the bulk rendering. This should cover major items like content blocks, question prompts and embedded components.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nTemplateConstants.java [hunk 1]
```diff
@@ -2,4 +2,7 @@
 
 public class I18nTemplateConstants {
     public static final String LAST_UPDATED_DATE_TEMPLATE_VAR_NAME = "LAST_UPDATED";
+    public static final String PARTICIPANT_FIRST_NAME = "DDP_PARTICIPANT_FIRST_NAME";
+    public static final String PARTICIPANT_LAST_NAME = "DDP_PARTICIPANT_LAST_NAME";
+    public static final String DASHED_DATE = "DDP_DASHED_DATE";
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nTemplateConstants.java:7]: Here are the new special vars we can drop into templates.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/FormInstance.java [hunk 2]
```diff
@@ -220,6 +225,26 @@ public void renderContent(Handle handle, I18nContentRenderer renderer, long lang
...
+                context.put(I18nTemplateConstants.PARTICIPANT_FIRST_NAME, profile.getFirstName());
+            }
+            if (profile.getLastName() != null) {
+                context.put(I18nTemplateConstants.PARTICIPANT_LAST_NAME, profile.getLastName());
+            }
+        }
+
+        return context;
+    }
+
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/FormInstance.java:247]: Here is where we fill in values for those special vars.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/MailingAddressComponent.java [hunk 3]
```diff
@@ -1,42 +1,77 @@
...
+            registry.accept(titleTemplateId);
+        }
+        if (subtitleTemplateId != null) {
+            registry.accept(subtitleTemplateId);
+        }
     }
 
     @Override
     public void applyRenderedTemplates(Provider<String> rendered, ContentStyle style) {
-        // Implement this method once we decide to use bulk rendering for this component
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/MailingAddressComponent.java:40]: Updated mailing-address component so it hooks into the bulk rendering process and can take advantage of those special vars.

File: study-builder/studies/testboston/snippets/consent-section4.conf [hunk 4]
```diff
@@ -95,6 +95,29 @@
...
+          <div>
+            <h3>Last Name</h3>
+            <p>$DDP_PARTICIPANT_LAST_NAME</p>
+          </div>
+          <div>
+            <h3>Date</h3>
+            <p>$DDP_DASHED_DATE</p>
+          </div>
+        """,
+        "variables": []
```
Comment (yufengwng) [inline study-builder/studies/testboston/snippets/consent-section4.conf:116]: Here is an example of activity definition using those special vars. Note that no variable objects are added to the `variables` list for the special vars.

**Review by charlottedye — APPROVED** [review]

**Review by MocanaAtBroad — APPROVED** [review]
Looks good. Does precisely what it needs to do. Nice!

**Comment by MocanaAtBroad** [comment]
Ok. Do we need to add templates (internationalized and with support for variables) in a bunch of other places? A crazy idea:  Could we look at centralizing the rendering of template strings and making the rendering annotation-driven? So in the case of the MailAddressComponent (not a good example because you have that inner class but let me try anyways, you would do something like:
```java
public class MailingAddressComponent extends FormComponent {
    @SerializedName("titleText")
    @Template("titleTemplateId")
    private String titleText;
    @SerializedName("subtitleText")
    @Template("subtitleTemplateId")
    private String subtitleText;
}
When time comes to serialize the object (perhaps from a route, perhaps, from the DAO), you calls something like:
```java
Renderer renderer = new Renderer(context); // context will know about the study, the participant the operator, etc
TemplateRenderer.render(mailAddressComponent);
```
And it would do all the template look ups and replacements.
So would help keep these model/like objects more like POJOs, not have to add special methods to each to do replacements, etc.

**Review by Sergei-Poliakov — APPROVED** [review]

**Pull request merged**